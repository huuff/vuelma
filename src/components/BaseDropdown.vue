<template>
<div class="dropdown" :class="{ 'is-active': open }">
  <div class="dropdown-trigger">
    <!-- TODO: More resilient aria-controls -->
    <button v-if="triggerText"
      class="button"
      aria-controls="dropdown-menu" 
      aria-haspopup="true"
      @click="open = !open"
      @blur="open = false"
      >
      {{ triggerText }}
    </button>
    <slot name="trigger"></slot>
  </div>
  <div class="dropdown-menu">
    <classed-slot :childrenClass="'dropdown-item'">
      <slot></slot>
    </classed-slot>
  </div>
</div>
</template>

<script setup lang="ts">
// TODO: Test it!
import { useSlots, ref } from 'vue';
import ClassedSlot from './ClassedSlot';

const props = defineProps<{
  triggerText?: string;
}>();

const open = ref(false);

const slots = useSlots();

if (props.triggerText && slots.trigger) {
  throw new Error("The dropdown component cannot be used with a trigger and trigger text at the same time!");
}
</script>
