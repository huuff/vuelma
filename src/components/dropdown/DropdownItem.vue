<template>
<slot></slot>
</template>

<script setup lang="ts">
import { OptionalId } from '@/util/optional-id';

export interface DropdownItemProps extends OptionalId {
  text: string;
  itemId?: string;
}

const props = defineProps<DropdownItemProps>();

</script>
