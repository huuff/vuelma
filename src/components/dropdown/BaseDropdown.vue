<template>
<div class="dropdown" :class="{ 'is-active': actualOpen }">
  <div class="dropdown-trigger">
    <button v-if="triggerText"
      class="button"
      aria-haspopup="true"
      @click="actualOpen = !actualOpen"
      >
      <template v-if="!$slots.trigger">
        <span> {{ triggerText }} </span>
        <span class="icon is-small">
          <font-awesome-icon icon="angle-down" />
        </span>
      </template>
      <template v-else>
        <slot name="trigger"></slot>
      </template>
    </button>
  </div>
  <div class="dropdown-menu">
    <dropdownContent></dropdownContent>
  </div>
</div>
</template>

<script setup lang="tsx">
// TODO: aria-controls
import { useSlots, toRef, } from 'vue';
import { useOptionalTwoWayBinding } from '@/composables/optional-two-way-binding';

import DropdownItem, { DropdownItemProps } from './DropdownItem.vue';
import DropdownDivider from './DropdownDivider.vue';


import partial from 'lodash/partial';
import classnames from 'classnames';

import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import { faAngleDown } from '@fortawesome/free-solid-svg-icons';
library.add(faAngleDown);

const props = withDefaults(defineProps<{
  triggerText?: string;
  open?: boolean;
  activeItemId?: string;
}>(), {
  open: undefined,
});

const emit = defineEmits<{
  (event: "update:open", value: boolean): void;
  (event: "update:activeItemId", value?: string): void;
}>();

// XXX: I don't know what's up with these issues, seems like
// the vue compiler went crazy. But it works anyway
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const actualOpen = useOptionalTwoWayBinding(false, toRef(props, "open"), partial(emit, "update:open"));
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
const actualActiveItemId = useOptionalTwoWayBinding(undefined, toRef(props, "activeItemId"), partial(emit, "update:activeItemId"));

const slots = useSlots();
if (!!props.triggerText === !!slots.trigger) {
  throw new Error("Only one of triggerText or the 'trigger' slot must be filled in!");
}

function isActive(itemId: string): boolean {
  return actualActiveItemId.value === itemId;
}

function setActive(itemId: string): void {
  actualActiveItemId.value = itemId;
  actualOpen.value = false;
}

function getItemId(props: DropdownItemProps) {
  return props.itemId ?? props.text;
}

function renderDropdownItem(itemProps: DropdownItemProps) {
  return <a 
          class={classnames({
            "dropdown-item": true,
            "is-active": isActive(getItemId(itemProps)),
          })}
          onClick={() => setActive(getItemId(itemProps))}
        >{itemProps.text}</a> 
}

const dropdownContent = () => 
<div class="dropdown-content">
  { slots.default && slots.default().map(el => {
    if (el.type === DropdownItem) 
      return renderDropdownItem(el.props as DropdownItemProps);
    else if (el.type === DropdownDivider)
      return el;
  })}
</div>
</script>
