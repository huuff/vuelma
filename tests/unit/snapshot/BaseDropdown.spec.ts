import { mount } from "@vue/test-utils";
import BaseDropdown from "@/components/dropdown/BaseDropdown.vue";
import DropdownItem from "@/components/dropdown/DropdownItem.vue";
import DropdownDivider from "@/components/dropdown/DropdownDivider.vue";

test("BaseDropdown.vue snapshot", () => {
  Math.random = jest.fn(() => 1);
  const wrapper = mount(BaseDropdown, {
    props: {
      triggerText: "Click here",
      activeItemId: "Item 1",
      color: "warning",
      open: true,
    }, 
    global: {
      components: { DropdownItem, DropdownDivider },
      stubs: [ "font-awesome-icon" ],
    },
    slots: {
      default: `
        <dropdown-item titleText="Item 1" />
        <div>Dropdown content</div>
        <dropdown-divider />
        <dropdown-item titleText="Item 2" />
      `,
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
