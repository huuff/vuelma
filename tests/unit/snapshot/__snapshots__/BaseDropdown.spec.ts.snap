// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseDropdown.vue snapshot 1`] = `
<div
  class="dropdown is-active"
>
  <div
    class="dropdown-trigger"
  >
    <button
      aria-controls="1"
      aria-haspopup="true"
      class="button is-warning"
    >
      
      <span>
        Click here
      </span>
      <span
        class="icon is-small"
      >
        <font-awesome-icon-stub
          border="false"
          fixedwidth="false"
          icon="angle-down"
          inverse="false"
          listitem="false"
          pulse="false"
          spin="false"
          swapopacity="false"
          symbol="false"
        />
      </span>
      
    </button>
  </div>
  <div
    class="dropdown-menu"
    id="id"
  >
    <div
      class="dropdown-content"
    >
      
      <a
        class="dropdown-item is-active"
        titletext="Item 1"
      >
        Item 1
      </a>
      <div
        class="dropdown-item"
      >
        <div>
          Dropdown content
        </div>
      </div>
      <hr
        class="dropdown-divider"
      />
      <a
        class="dropdown-item"
        titletext="Item 2"
      >
        Item 2
      </a>
      
    </div>
  </div>
</div>
`;
